storagenode/collector/service.go

```diff
@@ -1,33 +1,96 @@
 // Copyright (C) 2019 Storj Labs, Inc.
 // See LICENSE for copying information.
 
+// Package collector implements expired piece deletion from storage node.
 package collector
 
 import (
+	"context"
+	"errors"
 	"time"
 
 	"go.uber.org/zap"
+	monkit "gopkg.in/spacemonkeygo/monkit.v2"
 
+	"storj.io/storj/internal/sync2"
 	"storj.io/storj/storagenode/pieces"
 )
 
+var mon = monkit.Package()
+
 // Config defines parameters for storage node Collector.
 type Config struct {
-	Interval time.Duration
+	Interval time.Duration `help:"how frequently expired pieces are collected" default:"1h0m0s"`
 }
 
 // Service implements collecting expired pieces on the storage node.
 type Service struct {
 	log        *zap.Logger
 	pieces     *pieces.Store
 	pieceinfos pieces.DB
+
+	Loop sync2.Cycle
 }
 
 // NewService creates a new collector service.
-func NewService(log *zap.Logger, pieces *pieces.Store, pieceinfos pieces.DB) *Service {
+func NewService(log *zap.Logger, pieces *pieces.Store, pieceinfos pieces.DB, config Config) *Service {
 	return &Service{
 		log:        log,
 		pieces:     pieces,
 		pieceinfos: pieceinfos,
+		Loop:       *sync2.NewCycle(config.Interval),
 	}
 }
+
+// Run runs monitor service
+func (service *Service) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	return service.Loop.Run(ctx, func(ctx context.Context) error {
+		err := service.collect(ctx)
+		if err != nil {
+			service.log.Error("error during collecting pieces: ", zap.Error(err))
+		}
+		return nil
+	})
+}
+
+// Close stops the collector service.
+func (service *Service) Close() (err error) {
+	service.Loop.Close()
+	return nil
+}
+
+func (service *Service) collect(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	now := time.Now()
+	const maxBatches = 100
+	const batchSize = 1024
+
+	for k := 0; k < maxBatches; k++ {
```

Comment:
Can we take all of the code below outside of the loop and avoid having a max Batch Count?
Unlikely that we have to remove 1 Mio Entries within an hour, but still feeling wrong to me.

I guess you made that to avoid having two concurrent collect functions being run?

Reply:
Shouldn't we have the loop executing until `GetExpried` returns a non-empty result?

Reply:
We have both.

I'm not sure how to handle the situation where it is unable to delete the piece from disk (e.g. misplaced permissions). This protects against that issue. But there might be a cleaner way to handle this situation.

Reply:
What if the first 1024 (the batch limit) failed to delete? Then the loop will always get them again and again from the db and will never be able to delete other pieces that are ordered after them.

Maybe we should have a "delete_error" column in infodb and only try to delete those again only once per cycle.

Reply:
Yes, hence the reason there is the max batches... so it will stop after 100 tries.

Reply:
One option would be to delete from pieceinfo table and let user cleanup the situation.

But yes, we could add a some threshold for trying deletion.

Reply:
I am not sure if it is feasible for the SNO to manually delete piece by piece only those pieces that failed deletion. I would imagine the SNO to fix the problem (permissions or other) and then wait for the collector to execute once again.

Also, if the piece has already been deleted from the hard drive, trying to delete it once again won't result in an error, right?

Reply:
> Also, if the piece has already been deleted from the hard drive, trying to delete it once again won't result in an error, right?

Yes

Reply:
fixed